airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination/buffered_stream_consumer/BufferedStreamConsumer.java

```diff
@@ -148,8 +173,14 @@ protected void acceptTracked(final AirbyteMessage message) throws Exception {
 
   }
 
-  private void markStatesAsFlushedToTmpDestination() {
-    stateManager.markPendingAsFlushed();
+  /**
+   * After marking states as committed, emit the state message to platform then mark state messages as
+   * emitted to avoid resending the same state message to the platform
+   */
+  private void markStatesAsFlushedToDestination() {
```

Comment:
(nonblocking) one of the ideas brought up in the initial spec was that we'd do the flushes in a separate thread. This method is (probably?) not thread-safe, i.e. if there's a thread ingesting new record/state messages at the same that we're flushing stuff, weird things might happen

have you had a chance to sketch out what would need to change to support that sort of concurrency?

Reply:
I have not sketched out the changes needed to migrate this function to a thread-safe (atomic) operation. For the time being, that is something that would be a future consideration. There has been [discussion](https://airbytehq-team.slack.com/archives/C048E3BK1KM/p1674062757692439) on this topic, however, the goal was to first get a working checkpointing solution to not introduce any more scope creep work